### ISSUE TITLE
Bad compilation path if APK module depends on AAR module

### ISSUE THREAD

**Issue by tprochazka** [simpligility/android-maven-plugin#576]
If you have project with one APK module and several AAR modules used with main APK module and run **mvn package** command. Maven should build AAR modules first and then use it as dependency for the main APK module.

It almost works. Maven build modules first and then unpack final AAR to the target/unpacked-libs of the main APK module. But during compile task it didn't use then unpacked directory but directory with the name of the latest version in the repository.

For example if I have version 2.0.5-SNAPSHOT of my AAR library module it is unpacked to this folder
app/target/unpacked-libs/eial_android-about_2.0.5-SNAPSHOT
But compile task is looking for:
app/target/unpacked-libs/eial_android-about_2.0.5-20150112.101322-7/classes.jar
instead of
app/target/unpacked-libs/eial_android-about_2.0.5-SNAPSHOT/classes.jar

"2.0.5-20150112.101322-7" is latest version from my maven repo.

Related group thread
https://groups.google.com/forum/#!topic/maven-android-developers/AdLn3i2EJEw

**Comment by william-ferguson-au** [comment]
Great, so submit a pull request and I'll review it.


ᐧ

**Comment by tprochazka** [comment]
I know. But I am to busy now to study how it works, so maybe exist someone who knows exactly where  problem is and will be easy to fix it for them.

**Comment by kingargyle** [comment]
I'm running into this issue, as well.  If I build with a clean .m2 repository, the build tries to download the modules AAR file.   Once the build fails, I can do another build, and it will compile correctly.

@william-ferguson-au I'm willing to do a patch, but this is causing me production issues right now after upgrading to 4.1.1 and having to use AAR.  So I need something fairly quickly so that it picks up the AAR correctly that is part of the maven module build.

Otherwise I have to do some restructuring of our project get things fixed.

**Comment by william-ferguson-au** [comment]
@kingargyle your build will only fail if the AAR is not in your repo at the time it is required.

If you choose to do 'mvn package' as @tprochazka is doing then it will fail because your sub project AAR is not in you repo as it has not been installed yet.

If you use 'mvn install' then your sub project's AAR will be installed in your repo prior to it being required.

**Comment by kingargyle** [comment]
Actually that doesn't seem to be happening.   We have a SNAPSHOT of the AAR deployed from prior builds.  If we wipe out the local .m2 repository, have the AAR residing in the Nexus Repository.   It is being downloaded from Nexus, the AAR module is building with a mvn clean install, and the APK module builds, but fails with compilation issues.

If I immediately rerun the build, with mvn clean install without wiping out the repository.  The build will complete successfully.   I can repeat this exact same pattern on multiple machines.

I personally wouldn't expect the build to even try to download the AAR if the dependency is in the same build.  Maven correctly builds the AAR first before the APK, but the AAR doesn't seem to get added right.

I can try to put together a project that duplicates the issue.

**Comment by william-ferguson-au** [comment]
@kingargyle please create and publish a simple project that replicates and a new issue as what you are reporting is *not* what @tprochazka is talking about.

**Comment by kingargyle** [comment]
Well, I'll create a separate issue for the downloading of the AAR or dependencies for an APK that are in the module build.  But I am experiencing the same issue as @tprochazka as well, which is causing me to have to restructure our project to eliminate the embedded AAR module (which is fine as it was some left over legacy code).

**Comment by william-ferguson-au** [comment]
Why can't you use 'mvn install' instead of 'mvn package' to work around the issue?

**Comment by kingargyle** [comment]
We did use mvn install.  Always run the build with mvn clean install.  The problem is that we wipe out the workspace as part of every build.   So running with mvn clean install doesn't fix the issue for us.   The fact that it works the second time if you don't run mvn clean install, is scary...Which artifact is it picking up...the prior one, or the module that changed as part of the build.

**Comment by william-ferguson-au** [comment]
Create a new issue and provide as much info as possible. 
Optimally a cut down project that exposes the issue.

Cause what you are talking about is something quite different to @tprochazka

**Comment by kingargyle** [comment]
Ok will try and carve out some time this week for it.

**Comment by AlexeyTsiunchik** [comment]
I have also stuck with this issue. My project is multi module project which has aar library and few apk's which depends on aar. We periodically make CI builds and publish artifacts to the nexus. 
So initial state is: local repository is clean, remote nexus has snapshots previously built by Jenkins. I'm triggering build on parent pom (mvn clean install). 

While debugging I have noticed following behavior:
1.  After reading project configuration maven invokes 
```java
com.jayway.maven.plugins.android.phase_prebuild.ClasspathModifierLifecycleParticipant 
```
and it lists all modules and resolves dependencies.
```java 
line#126..... artifacts = dependencyResolver.getProjectDependenciesFor( project, session );
```
At this moment maven downloads my arr from nexus and place to the local repository artifact named global-core-4.0.0-20150219.111618-6.aar. Artifacts baseVersion is 4.0.0-SNAPSHOT, and version is  20150219.111618-6.
2. Next ClasspathModifierLifecycleParticipant adds reference to aar into classpath for all apks modules which depends on this common aar. 
```java
line#154 .... com.jayway.maven.plugins.android.phase_prebuild.ClasspathModifierLifecycleParticipant#addClassesToClasspath( helper, project, artifact );
```
2.1.  To compute classpath it invokes 
```java
com.jayway.maven.plugins.android.common.UnpackedLibHelper.getUnpackedClassesJar( Artifact artifact);
```
which compose classpath using code from line #170 and produces classpath entry like /somepath/unpacked-libs/ctc_global-core_20150219.111618-6/classes.jar
3. Next maven builds aar and install it into the local repo.
4. Next maven attempts to build my apks and resolve dependencies again on generate-sources phase 
```java
com.jayway.maven.plugins.android.phase01generatesources.GenerateSourcesMojo
line#525   final Collection<Artifact> artifacts = getTransitiveDependencyArtifacts(
                EXCLUDED_DEPENDENCY_SCOPES_FOR_EXTRACTION );
```
And it resolves my artifact to global-core-4.0.0-SNAPSHOT.aar which was built on previous step (both Artifact version and baseVersion are 4.0.0-SNAPSHOT). 
5.  It unpacks it into folder /somepath/unpacked-libs/ctc_global-core_4.0.0-SNAPSHOT/classes.jar.
6.  It executes maven compiler plugin which uses incorrect classpath entry generated on step 2.
7.  We got compilation error.

---------------------
Workaround:
1.  Use -o key while building project. With this key we prevent maven to download resources from remote repositories.  In this case on step 2 above it resolves artifact into global-core-4.0.0-SNAPSHOT.aar.
Issues:  it prevents downloading for other artifacts which may be not downloaded to your local repo yet.

**Comment by william-ferguson-au** [comment]
Excellent report @AlexeyTsiunchik! Much appreciated.

So it looks like the path for AARs in unpackedLibs (step 2.1 above) needs to be constructed from the base version for the artifact and no the actual version.

That shouldn't be too hard to fix.

**Comment by william-ferguson-au** [comment]
OK, so I can't replicate.

How or why are you getting a local artifact called `global-core-4.0.0-20150219.111618-6.aar` being downloaded from your Nexus repo?

Mine would download to local as global-core-4.0.0-SNAPSHOT.aar

**Comment by AlexeyTsiunchik** [comment]
In snapshot repository I can publish several Snapshots with the same version 4.0.0- SNAPSHOT. In this case internally nexus adds additional version to them see nexus screenshot ![Image of nexus](http://i57.tinypic.com/ofn821.png) .

If maven resolves such artifact from nexus it downloads them with version  4.0.0-xxxxxxx.xxxxxxx-xx  ![image of local repo] (http://i58.tinypic.com/2eqeur8.png) .

So after step 2.1 from my report local repo contains data as on image above.
And classpath entry is constructed based on this artifact.

**Comment by kingargyle** [comment]
Yep, seeing this same behavior as @AlexeyTsiunchik .  he beat me to giving you a detailed report.  Glad you guys can replicate the scenario.  Looking forward to a fix.

**Comment by william-ferguson-au** [comment]
@Manfred can you chime in. I cannot reproduce this.

I have a snaphsot repo on Nexus with multiple versions of my sub-module AAR.
When I create my project that includes my sub-module AAR, Maven doesn't
download the timestamped version of the  AAR SNAPSHOT, it just downloads
artifactId-version-SNAPSHOT.AAR

This is on Maven-3.2.5

Can you post a simple project so I can try to replicate on that?

ᐧ

**Comment by Manfred** [comment]
Wrong Manfred.

**Comment by william-ferguson-au** [comment]
:-) sorry @Manfred - should have been @mosabua

**Comment by AlexeyTsiunchik** [comment]
One more note, I'm executing mvn clean install on multimodule project which has parent module and few child modules which includes my aar and apks which depends on aar. I'm executing build on parent pom.![Image pom file] (http://i58.tinypic.com/2e31qo0.png)

**Comment by william-ferguson-au** [comment]
@AlexeyTsiunchik that sounds no different to my test project, so there must be something else.
Can you please cut down your project to just the AAR and one APK consuming it.
Gut the AAR and APK as much as possible.
And if it's still failing, then post the cutdown project.

**Comment by AlexeyTsiunchik** [comment]
I have created test project. Using local filesystem repository for deploy and having simple aar with interface and apk with class which implements this interface. Here is zip with sample https://www.dropbox.com/s/fmv3qxidd03aka2/sample.zip?dl=0

Steps:
1.  Unpack and execute mvn clean install (on parent pom). Expected results: compilation issue in APK module.
2.  Execute mvn clean install (on parent pom). Expected results: aar will be packaged and installed to local .M2 repo, APK still have compilation issue.
3.  Execute mvn clean install (on parent pom). Project compiled successfuly.

**Comment by william-ferguson-au** [comment]
I cannot reproduce this using your test project.
The results I get are that all 3 of them fail with:

```
[ERROR] Failed to execute goal on project global-core: Could not resolve dependencies for project by.alexts.common:global-core:aar:4.0.0-SNAPSHOT: The following
 artifacts could not be resolved: com.google.android:android:jar:4.2.2, com.google.android:support-v4:jar:18.0.1: Failure to find com.google.android:android:jar
:4.2.2 in http://Insomniac:8081/nexus/content/groups/public was cached in the local repository, resolution will not be reattempted until the update interval of
Nexus has elapsed or updates are forced -> [Help 1]
```

Which is 100% expected because AFAIK neither `com.google.android:android:jar:4.2.2` nor `com.google.android:support-v4:jar:18.0.1` exist

**Comment by jander234** [comment]
@william-ferguson-au I have the same problem. I have modified @AlexeyTsiunchik project removing unnecesary dependencies.

You can download it here: https://drive.google.com/file/d/0B_Om47NZHcUwa3RsQjZyQ1d0SDg/view?usp=sharing 

Steps:
1. Unpack and execute mvn clean install (on parent pom). Expected results: compilation issue in APK module.
2. Execute mvn clean install (on parent pom). Expected results: Project compiled successfuly.

**Comment by william-ferguson-au** [comment]
Yet again, I can't replicate. 
This is my output of that build using maven 3.2.3

```
d:\Temp\sample-android>mvn clean install
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] global-top
[INFO] global-core
[INFO] myapk
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building global-top 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ global-top ---
[INFO]
[INFO] --- maven-install-plugin:2.4:install (default-install) @ global-top ---
[INFO] Installing d:\Temp\sample-android\pom.xml to D:\dev\maven-local-repository\by\alexts\top\global-top\4.0.0-SNAPSHOT\global-top-4.0.0-SNAPSHOT.pom
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building global-core 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ global-core ---
[INFO] Deleting D:\Temp\sample-android\myaar\target
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myaar\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myaar\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myaar\target/filtered-assets
[INFO]
[INFO] --- android-maven-plugin:4.1.1:generate-sources (default-generate-sources) @ global-core ---
[INFO] Manifest copied from D:\Temp\sample-android\myaar\target\AndroidManifest.xml to D:\Temp\sample-android\myaar\target\AndroidManifest.xml
[INFO] Extracting libs
[INFO] Generating R file for by.alexts.common:global-core:aar:4.0.0-SNAPSHOT
[INFO] Adding R gen folder to compile classpath: D:\Temp\sample-android\myaar\target\generated-sources\r
[INFO] Adding AIDL gen folder to compile classpath: D:\Temp\sample-android\myaar\target\generated-sources\aidl
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myaar\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myaar\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myaar\target/filtered-assets
[INFO]
[INFO] --- maven-compiler-plugin:3.3:compile (default-compile) @ global-core ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to D:\Temp\sample-android\myaar\target\classes
[INFO]
[INFO] --- android-maven-plugin:4.1.1:proguard (default-proguard) @ global-core ---
[INFO] Proguard skipped because the configuration file doesn't exist: D:\Temp\sample-android\myaar\proguard.cfg
[INFO]
[INFO] --- maven-resources-plugin:2.5:testResources (default-testResources) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\Temp\sample-android\myaar\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.3:testCompile (default-testCompile) @ global-core ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.18:test (default-test) @ global-core ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.5:jar (default-jar) @ global-core ---
[INFO] Building jar: D:\Temp\sample-android\myaar\target\global-core.jar
[INFO]
[INFO] --- android-maven-plugin:4.1.1:aar (default-aar) @ global-core ---
[INFO] Generating AAR file : global-core
[INFO] Generating aar
[INFO] Building jar: D:\Temp\sample-android\myaar\target\global-core.aar.classes.jar
[INFO] D:\Temp\sample-android\myaar\src\main\libs does not exist, looking for libraries in target directory.
[INFO] Building zip: D:\Temp\sample-android\myaar\target\global-core.aar
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ global-core ---
[INFO] Installing D:\Temp\sample-android\myaar\target\global-core.aar to D:\dev\maven-local-repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\global-core-4
.0.0-SNAPSHOT.aar
[INFO] Installing D:\Temp\sample-android\myaar\pom.xml to D:\dev\maven-local-repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\global-core-4.0.0-SNAPSHOT.p
om
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building myapk 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ myapk ---
[INFO] Deleting D:\Temp\sample-android\myapk\target
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default) @ myapk ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myapk\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myapk\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myapk\target/filtered-assets
[INFO]
[INFO] --- android-maven-plugin:4.1.1:generate-sources (default-generate-sources) @ myapk ---
[INFO] Manifest copied from D:\Temp\sample-android\myapk\target\AndroidManifest.xml to D:\Temp\sample-android\myapk\target\AndroidManifest.xml
[INFO] Extracting libs
[INFO] Extracting aar global-core...
[INFO] Generating R file for by.alexts.myapk:myapk:apk:4.0.0-SNAPSHOT
[INFO] Generating R file for aar : by.alexts.core
[INFO] Adding R gen folder to compile classpath: D:\Temp\sample-android\myapk\target\generated-sources\r
[INFO] Skip BuildConfig.java generation for by.alexts.common global-core
[INFO] Adding AIDL gen folder to compile classpath: D:\Temp\sample-android\myapk\target\generated-sources\aidl
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ myapk ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myapk\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myapk\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myapk\target/filtered-assets
[INFO]
[INFO] --- maven-compiler-plugin:3.3:compile (default-compile) @ myapk ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 4 source files to D:\Temp\sample-android\myapk\target\classes
[INFO]
[INFO] --- android-maven-plugin:4.1.1:emma (default-emma) @ myapk ---
[INFO]
[INFO] --- android-maven-plugin:4.1.1:proguard (default-proguard) @ myapk ---
[INFO] Proguard skipped because the configuration file doesn't exist: D:\Temp\sample-android\myapk\proguard.cfg
[INFO]
[INFO] --- maven-resources-plugin:2.5:testResources (default-testResources) @ myapk ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\Temp\sample-android\myapk\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.3:testCompile (default-testCompile) @ myapk ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.18:test (default-test) @ myapk ---
[INFO] No tests to run.
[INFO]
[INFO] --- android-maven-plugin:4.1.1:dex (default-dex) @ myapk ---
[INFO] Error: The output jar is empty. Did you specify the proper '-keep' options?
[INFO] Exception in thread "main" java.lang.NullPointerException
[INFO]  at com.android.multidex.MainDexListBuilder.<init>(MainDexListBuilder.java:97)
[INFO]  at com.android.multidex.MainDexListBuilder.main(MainDexListBuilder.java:67)
[INFO] Convert classes to Dex : D:\Temp\sample-android\myapk\target
[INFO]
[INFO] --- maven-jar-plugin:2.5:jar (default-jar) @ myapk ---
[INFO] Building jar: D:\Temp\sample-android\myapk\target\myapk.jar
[INFO]
[INFO] --- android-maven-plugin:4.1.1:apk (default-apk) @ myapk ---
[INFO] Generating debug apk.
[INFO] Adding native libraries : []
[INFO]
[INFO] --- android-maven-plugin:4.1.1:internal-pre-integration-test (default-internal-pre-integration-test) @ myapk ---
[INFO] No InstrumentationRunner found - skipping tests
[INFO]
[INFO] --- android-maven-plugin:4.1.1:internal-integration-test (default-internal-integration-test) @ myapk ---
[INFO] No InstrumentationRunner found - skipping tests
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ myapk ---
[INFO] Installing D:\Temp\sample-android\myapk\target\myapk.apk to D:\dev\maven-local-repository\by\alexts\myapk\myapk\4.0.0-SNAPSHOT\myapk-4.0.0-SNAPSHOT.apk
[INFO] Installing D:\Temp\sample-android\myapk\pom.xml to D:\dev\maven-local-repository\by\alexts\myapk\myapk\4.0.0-SNAPSHOT\myapk-4.0.0-SNAPSHOT.pom
[INFO] Installing D:\Temp\sample-android\myapk\target\myapk.jar to D:\dev\maven-local-repository\by\alexts\myapk\myapk\4.0.0-SNAPSHOT\myapk-4.0.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] global-top ......................................... SUCCESS [  0.140 s]
[INFO] global-core ........................................ SUCCESS [  2.758 s]
[INFO] myapk .............................................. SUCCESS [  2.261 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.914 s
[INFO] Finished at: 2015-06-11T17:25:06+10:00
[INFO] Final Memory: 29M/698M
[INFO] ------------------------------------------------------------------------
d:\Temp\sample-android>
```

**Comment by jander234** [comment]
@william-ferguson-au Delete 'by' folder from your .m2 local repository and try again.

**Comment by william-ferguson-au** [comment]
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] global-top ......................................... SUCCESS [  0.138 s]
[INFO] global-core ........................................ SUCCESS [  3.014 s]
[INFO] myapk .............................................. SUCCESS [  2.222 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.433 s
[INFO] Finished at: 2015-06-11T17:57:45+10:00
[INFO] Final Memory: 31M/707M
[INFO] ------------------------------------------------------------------------
d:\Temp\sample-android>
```

**Comment by william-ferguson-au** [comment]
Post your full build output.
How is it failing *exactly*.

**Comment by jander234** [comment]
Here my first full build output. Note the downloading phase after 'Scanning for projects...':

```
[INFO] Scanning for projects...
Downloading: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml
777/777 B   
            
Downloaded: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml (777 B at 2.6 KB/sec)
Downloading: http://ci:8080/repository/snapshots/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml
            
Downloading: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-20150226.160450-5.aar
4/32 KB     
8/32 KB   
12/32 KB   
16/32 KB   
20/32 KB   
24/32 KB   
28/32 KB   
32/32 KB   
           
Downloaded: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-20150226.160450-5.aar (32 KB at 1052.5 KB/sec)
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] global-top
[INFO] global-core
[INFO] myapk
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building global-top 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ global-top ---
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ global-top ---
[INFO] Installing D:\Temp\sample-android\pom.xml to C:\Users\jahernandez\.m2\repository\by\alexts\top\global-top\4.0.0-SNAPSHOT\global-top-4.0.0-SNAPSHOT.pom
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building global-core 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ global-core ---
[INFO] Deleting D:\Temp\sample-android\myaar\target
[INFO] 
[INFO] --- maven-resources-plugin:2.5:resources (default) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myaar\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myaar\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myaar\target/filtered-assets
[INFO] 
[INFO] --- android-maven-plugin:4.1.1:generate-sources (default-generate-sources) @ global-core ---
[INFO] Manifest copied from D:\Temp\sample-android\myaar\target\AndroidManifest.xml to D:\Temp\sample-android\myaar\target\AndroidManifest.xml
[INFO] Extracting libs
[INFO] Generating R file for by.alexts.common:global-core:aar:4.0.0-SNAPSHOT
[INFO] Adding R gen folder to compile classpath: D:\Temp\sample-android\myaar\target\generated-sources\r
[INFO] Adding AIDL gen folder to compile classpath: D:\Temp\sample-android\myaar\target\generated-sources\aidl
[INFO] 
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myaar\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myaar\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myaar\target/filtered-assets
[INFO] 
[INFO] --- maven-compiler-plugin:3.3:compile (default-compile) @ global-core ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to D:\Temp\sample-android\myaar\target\classes
[INFO] 
[INFO] --- android-maven-plugin:4.1.1:proguard (default-proguard) @ global-core ---
[INFO] Proguard skipped because the configuration file doesn't exist: D:\Temp\sample-android\myaar\proguard.cfg
[INFO] 
[INFO] --- maven-resources-plugin:2.5:testResources (default-testResources) @ global-core ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\Temp\sample-android\myaar\src\test\resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.3:testCompile (default-testCompile) @ global-core ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.17:test (default-test) @ global-core ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:2.5:jar (default-jar) @ global-core ---
[INFO] Building jar: D:\Temp\sample-android\myaar\target\global-core.jar
[INFO] 
[INFO] --- android-maven-plugin:4.1.1:aar (default-aar) @ global-core ---
[INFO] Generating AAR file : global-core
[INFO] Generating aar
[INFO] Building jar: D:\Temp\sample-android\myaar\target\global-core.aar.classes.jar
[INFO] D:\Temp\sample-android\myaar\src\main\libs does not exist, looking for libraries in target directory.
[INFO] Building zip: D:\Temp\sample-android\myaar\target\global-core.aar
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ global-core ---
[INFO] Installing D:\Temp\sample-android\myaar\target\global-core.aar to C:\Users\jahernandez\.m2\repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\global-core-4.0.0-SNAPSHOT.aar
[INFO] Installing D:\Temp\sample-android\myaar\pom.xml to C:\Users\jahernandez\.m2\repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\global-core-4.0.0-SNAPSHOT.pom
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building myapk 4.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ myapk ---
[INFO] Deleting D:\Temp\sample-android\myapk\target
[INFO] 
[INFO] --- maven-resources-plugin:2.5:resources (default) @ myapk ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myapk\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myapk\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myapk\target/filtered-assets
[INFO] 
[INFO] --- android-maven-plugin:4.1.1:generate-sources (default-generate-sources) @ myapk ---
[INFO] Manifest copied from D:\Temp\sample-android\myapk\target\AndroidManifest.xml to D:\Temp\sample-android\myapk\target\AndroidManifest.xml
[INFO] Extracting libs
[INFO] Extracting aar global-core...
[INFO] Generating R file for by.alexts.myapk:myapk:apk:4.0.0-SNAPSHOT
[INFO] Generating R file for aar : by.alexts.core
[INFO] Adding R gen folder to compile classpath: D:\Temp\sample-android\myapk\target\generated-sources\r
[INFO] Skip BuildConfig.java generation for by.alexts.common global-core
[INFO] Adding AIDL gen folder to compile classpath: D:\Temp\sample-android\myapk\target\generated-sources\aidl
[INFO] 
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ myapk ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 11 resources to D:\Temp\sample-android\myapk\target/filtered-res
[INFO] Copying 1 resource to D:\Temp\sample-android\myapk\target
[INFO] Copying 0 resource to D:\Temp\sample-android\myapk\target/filtered-assets
[INFO] 
[INFO] --- maven-compiler-plugin:3.3:compile (default-compile) @ myapk ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 4 source files to D:\Temp\sample-android\myapk\target\classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /D:/Temp/sample-android/myapk/src/by/alexts/core/MyClass.java:[3,33] cannot find symbol
  symbol: class MyInterface
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] global-top ......................................... SUCCESS [  0.623 s]
[INFO] global-core ........................................ SUCCESS [  3.410 s]
[INFO] myapk .............................................. FAILURE [  1.420 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 8.276 s
[INFO] Finished at: 2015-06-11T10:09:14+02:00
[INFO] Final Memory: 28M/248M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project myapk: Compilation failure
[ERROR] /D:/Temp/sample-android/myapk/src/by/alexts/core/MyClass.java:[3,33] cannot find symbol
[ERROR] symbol: class MyInterface
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :myapk
```

**Comment by AlexeyTsiunchik** [comment]
This is how I finally manage to resolve it locally ![local fix] (http://i62.tinypic.com/6xqlar.png) It is rude fix, but finally it makes our builds working. I'm not sure which side effects it may have so I have not proposed it as pull request.

**Comment by mosabua** [comment]
That might be fine as a fix to be honest. But I would like to understand first why it doesnt work for you when it works for @william-ferguson-au  ... can you try with the same Maven version?

**Comment by kingargyle** [comment]
I'm wondering if it has to do with the maven version in use.  I see this as well but I'm running maven 3.1.1, not sure what @AlexeyTsiunchik is running version wise.

**Comment by william-ferguson-au** [comment]
I need to retest. While I removed the artifacts from my local repo pretest,
it seems they are in my Nexus repo and are being pulled from there.

**Comment by william-ferguson-au** [comment]
OK scrap that. I don't know what made me think it was retrieving the artifacts from Nexus cause it wasn't and isn't.

I cannot make this fail either for Maven 3.2.3 or 3.1.1

What config do you have in your Maven Settings.xml?

**Comment by AlexeyTsiunchik** [comment]
Hello William, here is info.
1. Maven version.
```shell
C:\Users\Alexey>mvn --version
Apache Maven 3.2.2 (45f7c06d68e745d05611f7fd14efb6594181933e; 2014-06-17T16:51:42+03:00)
Maven home: d:\dev_tools\apache-maven-3.2.2
Java version: 1.7.0_60, vendor: Oracle Corporation
Java home: C:\Program Files (x86)\Java\jdk1.7.0_60\jre
Default locale: en_US, platform encoding: Cp1251
OS name: "windows 8.1", version: "6.3", arch: "x86", family: "windows"
```
2. Settings.xml has been removed from repo folder. In Maven home dir it is unchanged.
3. Steps to reproduce.
3.1. Download my archive and unpack it.
3.2  Remove $M2_REPO/by/alexts dir (remove artifacts from your local repository)
3.3  Execute mvn clean package (on parent pom). Expected results: compilation issue in APK module.
3.4. Execute mvn clean install (on parent pom). Expected results: aar will be packaged and installed to your local .M2 repo, APK still have compilation issue.
3.5. Execute mvn clean install (on parent pom). Project compiled successfuly.

**Comment by william-ferguson-au** [comment]
Sorry @AlexeyTsiunchik I cannot make this fail.

I removed my Settings.xml.
I removed the 'by' folder from my local repo.
'mvn clean package' works fine.
'mvn clean install' works fine.
'mvn clean install' works fine.

Everything built OK.

The only strange occurrence was the following output during 'mvn clean package' and 'mvn clean install' when building the my-apk module. But it still created an apk for me.
```
[INFO] --- android-maven-plugin:4.1.1:dex (default-dex) @ myapk ---
[INFO] Error: The output jar is empty. Did you specify the proper '-keep' options?
[INFO] Exception in thread "main" java.lang.NullPointerException
[INFO]  at com.android.multidex.MainDexListBuilder.<init>(MainDexListBuilder.java:97)
[INFO]  at com.android.multidex.MainDexListBuilder.main(MainDexListBuilder.java:67)
[INFO] Convert classes to Dex : D:\Temp\sample-android\myapk\target
```

**Comment by AlexeyTsiunchik** [comment]
@william-ferguson-au Here are my thoughts:
1. Are you using my original package from https://www.dropbox.com/s/fmv3qxidd03aka2/sample.zip?dl=0 ? Possibly you have changed something. Could you please try it again on original one?
2. Are your version of android-maven-plugin is the same as mine? Mine has been downloaded from maven-central, and has version 4.1.1.
3.  Possibly you are executing other command (mvn clean install instead on mvn clean package)?

Please find output from mvn clean package -X here (https://www.dropbox.com/s/zjmuhw41vyo6c1g/out.zip?dl=0). Could you execute same command on your environment and share output ?

**Comment by william-ferguson-au** [comment]
@AlexeyTsiunchik I am not using your project from 3 months ago, because as I noted back then it is using dependencies that don't exist so it could never have built.

I am using the project that @jander234 provided which cut out the non-existing deps.

I am using the version of the android-maven-plugin defined in the project.

I am executing 'mvn clean package'

**Comment by AlexeyTsiunchik** [comment]
@william-ferguson-au  I have tried on project posted by @jander234, it is the same. Output is here: https://www.dropbox.com/s/fmv3qxidd03aka2/sample.zip?dl=0 . Could you please run mvn clean package -X and share output ?

**Comment by william-ferguson-au** [comment]
@AlexeyTsiunchik the link you provided is for your project from 3 months ago. It has no build output.

mvn clean package -X
Output at https://www.dropbox.com/s/tclf2y7pipp6wp2/build-out.txt?dl=0

**Comment by AlexeyTsiunchik** [comment]
@william-ferguson-au William, please pay attention on line #835 of your output file.
Downloading: http://Insomniac:8081/nexus/content/groups/public/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-SNAPSHOT.aar

Seems you are using Nexus repository and my artifact was installed there. Due to this it is always downloaded as 4.0.0-SNAPSHOT.aar and you do not see this problem.

Next interesting line is #842 
[DEBUG] Adding to classpath : D:\Temp\sample-android\myapk\target\unpacked-libs\bac_global-core_4.0.0-SNAPSHOT\classes.jar

Correspondent line in my output is #262 
[DEBUG] Adding to classpath : D:\projects\sample\sample-android\myapk\target\unpacked-libs\bac_global-core_4.0.0-20150226.160450-5\classes.jar

Could you please try to build sample project without nexus?

**Comment by william-ferguson-au** [comment]
You artifacts is **NOT** in my nexus repo.

Maven *attempts* to download it from Nexus by fails.
Hitting that URL up gives:
```
404 - Path
/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-SNAPSHOT.aar
not found in group repository "Public Repositories" [id=public].

Path
/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-SNAPSHOT.aar
not found in group repository "Public Repositories" [id=public].
```

As noted above I have already tried building sans Nexus.
But to prove it to you here is the build output *without* Nexus.
https://www.dropbox.com/s/a9tq3sryai0zrn8/build-out-sans-nexus.txt?dl=0

ᐧ

**Comment by AlexeyTsiunchik** [comment]
Hello @william-ferguson-au, seems I finally find differences.

For some reason in sample project you have no D:\Temp\sample-android\local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml file.

lines 638-640 from your output
````log 
[DEBUG] Could not find metadata by.alexts.common:global-core:4.0.0-SNAPSHOT/maven-metadata.xml in local (D:\dev\maven-local-repository)
[DEBUG] Using connector WagonRepositoryConnector with priority 0.0 for file://D:\Temp\sample-android\myapk/../local-repo
Downloading: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml
````

In mine output you may notice
````log
[DEBUG] Could not find metadata by.alexts.common:global-core:4.0.0-SNAPSHOT/maven-metadata.xml in local (C:\Users\Alexey\.m2\repository)
[DEBUG] Using connector WagonRepositoryConnector with priority 0.0 for file://D:\projects\sample\sample-android\myapk/../local-repo
Downloading: file://D:\projects\sample\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml
777/777 B    
            
Downloaded: file://D:\projects\sample\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml (777 B at 14.3 KB/sec)
[DEBUG] Writing tracking file C:\Users\Alexey\.m2\repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\resolver-status.properties
````

Without this file maven didn't notice different SNAPSHOTS versions and download jar with version 4.0.0-SNAPSHOT
````log
[DEBUG] Failure to find by.alexts.common:global-core:4.0.0-SNAPSHOT/maven-metadata.xml in file://D:\Temp\sample-android\myapk/../local-repo was cached in the local repository, resolution will not be reattempted until the update interval of local-repo has elapsed or updates are forced
[DEBUG] Using connector WagonRepositoryConnector with priority 0.0 for file://D:\Temp\sample-android\myapk/../local-repo
Downloading: file://D:\Temp\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-SNAPSHOT.aar
[DEBUG] Writing tracking file D:\dev\maven-local-repository\by\alexts\common\global-core\4.0.0-SNAPSHOT\global-core-4.0.0-SNAPSHOT.aar.lastUpdated
````

If maven-metadata is available it download artifact with minor version
````log
[DEBUG] Could not find metadata by.alexts.common:global-core:4.0.0-SNAPSHOT/maven-metadata.xml in local (C:\Users\Alexey\.m2\repository)
[DEBUG] Skipped remote update check for by.alexts.common:global-core:4.0.0-SNAPSHOT/maven-metadata.xml, already updated during this session.
[DEBUG] Using connector WagonRepositoryConnector with priority 0.0 for file://D:\projects\sample\sample-android\myapk/../local-repo
Downloading: file://D:\projects\sample\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-20150226.160450-5.aar
4/32 KB     
8/32 KB   
12/32 KB   
16/32 KB   
20/32 KB   
24/32 KB   
28/32 KB   
32/32 KB   
           
Downloaded: file://D:\projects\sample\sample-android\myapk/../local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/global-core-4.0.0-20150226.160450-5.aar (32 KB at 1503.6 KB/sec)
````

** Note **: If I remove /local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml project compiled successfully.

**Comment by william-ferguson-au** [comment]
Right ..
I removed all instances of any /by/alexts/ artifacts from my local maven
repo.
I also removed them from the repo which seemed to have been attached to the
project.

That was the point right?
That you can build this project from scratch.

Since the POM had defined a local-repo (NB this is generally a very bad
idea) they needed to be removed from there too.

So you are all good now?

ᐧ

**Comment by AlexeyTsiunchik** [comment]
@william-ferguson-au, well this is not exactly what I want to say. This file is automatically created by maven while uploading artifact in repository using mvn deploy. So while you are using any kind of repository which manage SNAPSHOT artifacts (for example, for continues integration) you will encounter this issue. I was using local repository only for demonstration purposes, to simulate remote nexus repository with artifacts. We experience the same issue while using nexus for continues integration. 

I have resolved our issue by applying patch and build custom version of the plugin. I have added my posts only to point you with exact root cause of issue.

**Comment by william-ferguson-au** [comment]
So why don't I get it when using my nexus repo?


ᐧ

**Comment by william-ferguson-au** [comment]
Or if I execute 'mvn deploy' on the sample project (which will deploy to
//local-repo) and then 'mvn clean package', why does it still build for me?
Isn't this the exact use case that is failing for you?


ᐧ

**Comment by AlexeyTsiunchik** [comment]
Answering on your first question: you can get it with nexus but need more steps.
1.  Update pom and set snapshot repository to your nexus repo url.
2.  Execute mvn package deploy 2-3 times to upload 3 different artifacts to your remote nexus repository.
3.  Add some method to interface in aar and some implementation in apk module.
4.  Remove artifacts from M2_REPO/by/alexts 
5.  Execute mvn clean package you should get compilation error.

Since this path is longer I have created local repo and demonstrate issue based on it.

**Comment by AlexeyTsiunchik** [comment]
For second comment:
1. Please execute mvn clean deploy
2. Clear $M2_REPO\by\ folder
3. Execute mvn clean package - you will get compilation exception

**Comment by william-ferguson-au** [comment]
OK well I shouldn't need to do (1) because the local-repo is marked as snapshot, so delpoy will go to there.
2) Deploys to local-repo
3) why is this required?
4) Is done - as we have with all the tests
5) No I don't get the compilation error

And for the 2nd comment:
When I follow those 3 steps - again I **don't** get the compilation failure

**Comment by AlexeyTsiunchik** [comment]
Do you have /local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml  after execution mvn clean deploy?

**Comment by william-ferguson-au** [comment]
@mosabua can you spin it up. I'm obviously missing something obvious.

**Comment by william-ferguson-au** [comment]
@AlexeyTsiunchik yes - after 'mvn clean deploy' /local-repo/by/alexts/common/global-core/4.0.0-SNAPSHOT/maven-metadata.xml exists

**Comment by AlexeyTsiunchik** [comment]
Please execute mvn deploy again to have 2 different SNAPSHOT artifacts in repo, 
and next remove D:\dev\maven-local-repository\by folder
mvn clean package generates compilation issue

**Please note**: you should remove D:\dev\maven-local-repository\by not D:\Temp\sample-android/local-repo/by

**Comment by william-ferguson-au** [comment]
@AlexeyTsiunchik
- I executed 'mvn clean deploy' 3 times like you suggested previously.
- Removed `D:\dev\maven-local-repository\by`
- Executed 'mvn clean package`

I **do not** get a compilation issue

**Comment by mosabua** [comment]
This project does not build for me. I get a problem with the dex command in the apk module. 

[INFO] global-core ........................................ SUCCESS [  4.171 s]
[INFO] myapk .............................................. FAILURE [  1.694 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 8.290 s
[INFO] Finished at: 2015-06-15T21:39:52-07:00
[INFO] Final Memory: 28M/322M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.simpligility.maven.plugins:android-maven-plugin:4.1.1:dex (default-dex) on project myapk: Failed to execute mainDexClasses: ANDROID-040-001: Could not execute: Command = /bin/sh -c cd /Users/manfred/Downloads/sample-android/myapk && /Volumes/lin-data/tools/android-sdk-mac_86/build-tools/22.0.1/mainDexClasses --output /Users/manfred/Downloads/sample-android/myapk/target/mainDexClasses.txt "/Users/manfred/Downloads/sample-android/myapk/target/classes:/Users/manfred/Downloads/sample-android/myapk/target/unpacked-libs/bac_global-core_4.0.0-SNAPSHOT/classes.jar", Result = 10 -> [Help 1]

**Comment by mosabua** [comment]
This might be due some part of the project setup and me running it on OS X.

**Comment by AlexeyTsiunchik** [comment]
Ok, this is my last attempt to demonstrate this issue. Please check video on [youtube](https://www.youtube.com/watch?v=b9oO8yw4JSE)

**Comment by Shusshu** [comment]
You guys should really be using a project hosted via git instead of sharing random zip files....

This is my output http://pastebin.com/0ydwBCR8 based on the project  : https://drive.google.com/file/d/0B_Om47NZHcUwa3RsQjZyQ1d0SDg/view?usp=sharing

**Comment by mosabua** [comment]
I agree with the fact that this issue occurs for you. I however do not have time to look at this until probably September. The problem probably lies with the lifecycle participant and how it resolves the aar from the repo and puts it on the classpath (fails to do so in fact).

It seems also weird that it works for @william-ferguson-au but fails for @AlexeyTsiunchik and @Shusshu and beyond that I am kind of thinking that it should resolve it from within the reactor and work instead of getting it from the repo and fail.

**Comment by aleksei-pavlenko** [comment]
We are experiencing similar problems. Do you have time to look at it?

**Comment by nbeuret** [comment]
We also had the same issue on our CI since we have updated the plugin to the 4.0+ version (we are currently using the 4.3.0 release).

To workaround the issue, we have deployed on our internal plugin repository a snapshot of the 4.3.0 release with the fix of @AlexeyTsiunchik. It's working perfectly. It would be nice if the fix is part of the next release to get rid of this workaround.

**Comment by Stuey86** [comment]
Really Glad I finally found this issue.
We are also getting the same issue on Projects that contain Multiple AAR Modules that Depend on Eachother. 
We can solve the issue by disabling repo (Nexus) snapshots & only use local snapshots - Not a proper solution.
All our projects that that don't depend on interanal AAR modules build correctly every-time.
`...`
`    <snapshots>`
`        <enabled>false</enabled>`
`    </snapshots>`
`</repository>`
The issue really does seem to be related to time-stamps on SNAPSHOTs when using a Repo like Nexus.

When repo snapshots are enabled the First build Fails but Second Passes.
2.	Delete folder .m2\repository\com\example\myProject
3.	Mvn clean install -- Fails
4.	Mvn clean install -- Works
5.	Delete folder  .m2\repository\com\example\myProject
6.	Mvn clean install -- Fails
7.	Mvn clean install -- Works
8.     Rinse and Repeat.

Really does smell like something to do with Time-Stamps.

**Comment by mkontio** [comment]
We also have the same problem - hope that this would get fixed soon!

**Comment by Stuey86** [comment]
Fix from @AlexeyTsiunchik works for us as well.

**Comment by Stuey86** [comment]
Hi @mosabua. Will anything be done for this on the next release? Thanks

**Comment by william-ferguson-au** [comment]
So does someone who can replicate the failure want to submit a pull request of the fix provided by @AlexeyTsiunchik ?

**Comment by nbeuret** [comment]
Actually it looks like the fix has already been merged with the commit 
https://github.com/simpligility/android-maven-plugin/commit/ac03df5c3344c540066867d5b1cba8f87742e167 from the PR https://github.com/simpligility/android-maven-plugin/pull/671

**Comment by william-ferguson-au** [comment]
Excellent.
